File: src/GaussianFilter.java
package pip;

import java.awt.Color;
import java.awt.image.BufferedImage;

/**
 *
 * @author Ovilia
 */
public class GaussianFilter extends Filter {

    double sigma;

    public GaussianFilter(BufferedImage originImage, int radio, double sigma,
            BorderProcess borderProcess) {
        super(originImage, radio, borderProcess);
        this.sigma = sigma;
    }

    protected void createFilteredImage() {
        double coefficient = 0.5 / Math.PI / sigma / sigma;
        double[][] kernel = new double[2 * radio + 1][2 * radio + 1];
        double kernelSum = 0;
        for (int m = 0; m <= 2 * radio; ++m) {
            for (int n = 0; n <= 2 * radio; ++n) {
                kernel[m][n] = coefficient * Math.exp(
                        -(m * m + n * n) / 2 / sigma / sigma);
                kernelSum += kernel[m][n];
                //System.out.println(kernel[m][n]);
            }
        }
        // make sure brightness doesn't change
        for (int m = 0; m <= 2 * radio; ++m) {
            for (int n = 0; n <= 2 * radio; ++n) {
                kernel[m][n] /= kernelSum;
            }
        }

        int width = filteredImage.getWidth();
        int height = filteredImage.getHeight();
        for (int i = 0; i < width; ++i) {
            for (int j = 0; j < height; ++j) {
                double value = 0;
                for (int m = 0; m <= 2 * radio; ++m) {
                    for (int n = 0; n <= 2 * radio; ++n) {
                        int gray = (new Color(borderedImage.getRGB(i + m, j + n))).getRed();
                        value += gray * kernel[m][n];
                    }
                }
                int color = (int) value;
                if (color > ImageProcessor.RANGE_OF_8BITS - 1) {
                    color = ImageProcessor.RANGE_OF_8BITS - 1;
                }
                filteredImage.setRGB(i, j, (new Color(color, color, color)).getRGB());
            }
        }
    }
}

File: src/GaussianFilter.cpp
#include <cmath>

#include "GaussianFilter.h"
#include "ImagePolicy.h"

GaussianFilter::GaussianFilter(QImage* image, const double sigma,
                               const bool isColored,
                               ImagePolicy::BorderPolicy policy) :
    // if radio is less than 1, set it to be 1
    LinearFilter(image, (int)(sigma * RADIO_DIVIDED_BY_SIGMA) < 1 ?
                     1 : (int)(sigma * RADIO_DIVIDED_BY_SIGMA),
                 isColored, policy),
    kernelLength((2 * kernelRadio + 1) * (2 * kernelRadio + 1)),
    sigma(sigma)
{
    kernelPtr = new double[kernelLength];

    // init kernel
    double coefficient = 0.5 / M_PI / sigma / sigma;
    double sum = 0;
    int index = 0;
    for (int i = -kernelRadio; i <= kernelRadio; ++i) {
        for (int j = -kernelRadio; j <= kernelRadio; ++j) {
            double kernel = coefficient *
                    exp((i * i + j * j) / (-2) / sigma / sigma);
            kernelPtr[index] = kernel;
            ++index;
            sum += kernel;
        }
    }
    // make sure sum is 1
    for (int i = 0; i < kernelLength; ++i) {
        kernelPtr[i] /= sum;
    }

    LinearFilter::changeKernel(kernelPtr);
}

GaussianFilter::~GaussianFilter()
{
    delete kernelPtr;
}

double GaussianFilter::getSigma()
{
    return sigma;
}
